libbpf-tools/tcpretrans.bpf.c

```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Anton Protopopov
+// Copyright (c) 2021 Red Hat
+//
+// Based on tcpconnect.c by Anton Protopopov and
+// tcpretrans(8) from BCC by Brendan Gregg
+#include <vmlinux.h>
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+
+#include "maps.bpf.h"
+#include "tcpretrans.h"
+
+/* Define here, because there are conflicts with include files */
+#define AF_INET		2
+#define AF_INET6	10
+
+SEC(".rodata") const volatile bool do_count = 0;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct ipv4_flow_key);
+	__type(value, u64);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} ipv4_count SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct ipv6_flow_key);
+	__type(value, u64);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} ipv6_count SEC(".maps");
+
+static  __always_inline void count_v4(const struct sock *skp)
+{
+	struct ipv4_flow_key key = {};
+	static __u64 zero;
+	__u64 *val;
+
+	BPF_CORE_READ_INTO(&key.saddr, skp, __sk_common.skc_rcv_saddr);
+	BPF_CORE_READ_INTO(&key.daddr, skp, __sk_common.skc_daddr);
+	BPF_CORE_READ_INTO(&key.dport, skp, __sk_common.skc_dport);
+	BPF_CORE_READ_INTO(&key.sport, skp, __sk_common.skc_num);
+	val = bpf_map_lookup_or_try_init(&ipv4_count, &key, &zero);
+	if (val)
+		__atomic_add_fetch(val, 1, __ATOMIC_RELAXED);
+}
+
+static __always_inline void count_v6(const struct sock *skp)
+{
+	struct ipv6_flow_key key = {};
+	static const __u64 zero;
+	__u64 *val;
+
+	BPF_CORE_READ_INTO(&key.saddr, skp,
+			   __sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
+	BPF_CORE_READ_INTO(&key.daddr, skp,
+			   __sk_common.skc_v6_daddr.in6_u.u6_addr32);
+	BPF_CORE_READ_INTO(&key.dport, skp, __sk_common.skc_dport);
+	BPF_CORE_READ_INTO(&key.sport, skp, __sk_common.skc_num);
+
+	val = bpf_map_lookup_or_try_init(&ipv6_count, &key, &zero);
+	if (val)
+		__atomic_add_fetch(val, 1, __ATOMIC_RELAXED);
+}
+
+static int trace_event(void *ctx, const struct sock *skp, int type)
+{
+	if (skp == NULL)
+		return 0;
+
+	struct event e = {};
+	__u16 dport;
+	__u16 sport;
+	__u32 family;
+	__u64 pid_tgid;
+	__u32 pid;
+	int state;
+
+	family = BPF_CORE_READ(skp, __sk_common.skc_family);
+	e.af = family;
+
+	if (do_count) {
+		if (family == AF_INET)
+			count_v4(skp);
+		else
+			count_v6(skp);
+		return 0;
+	}
+
+	e.type = type;
+	pid_tgid = bpf_get_current_pid_tgid();
+	pid = pid_tgid >> 32;
+	e.pid = pid;
+
+	BPF_CORE_READ_INTO(&dport, skp, __sk_common.skc_dport);
+	e.dport = dport;
+	BPF_CORE_READ_INTO(&sport, skp, __sk_common.skc_num);
+	e.sport = sport;
+	state = BPF_CORE_READ(skp, __sk_common.skc_state);
+	e.state = state;
```

Comment:
Though no big deal, I think you can save many of these local variables.

Reply:
e.state complains about type casting, I'll remove the port variables, they are unneeded.